docs: use vitepress

marcalexiei opened the pull request:
## Description

Use `vitepress`to build doc.

## Motivation and Context

See associated issue: #3905

## Usage examples

n/a

## TODOs

- [x] Convert all docs
  - [x] Guides
  - [x] Concepts
  - [x] Reference
  - [x] Improve homepage
- [x] Update documentation unit tests
- [x] ~~Add copyright in the footer?~~ [not needed](https://github.com/conventional-changelog/commitlint/pull/3921#issuecomment-1951339113)
- [x] Be sure to run `prettier` on docs files when necessary
- [x] Update CONTRIBUTING.md with new docs commands
- [x] Automatic install via CI
- [x] Replace `module.exports` in examples with ESM examples

## Nav changes

### Guides
- "Release" and "Upgrade commitlint" have been moved inside  "Support"

- new "Getting started" page inside Guides (contains shared item for "Local setup" and "CI setup")

### References

Rules page has been divided into two pages:
- Rules: contains the list of all rules
- Rules configuration: contains all the information about the rule (section on the top of the old rule page)

### API

New section with a single page for each package (format, lint, load and read)

### Support

a new collapsible section named "Support" which contains section.
In the future will include more content like troubleshooting information (#3894)

### "Attributions" 

new section with previous "Attributions" from docs/README.md content

## Page preview (WIP)

<details>
<summary>

### Home

</summary>

<img width="1278" alt="image" src="https://github.com/conventional-changelog/commitlint/assets/24919330/618b59eb-1bf8-438b-bfde-5374d35cf7af">

</details>

<details>
<summary>

### Guides: Local setup

</summary>

![Screenshot 2024-02-15 at 20 45 21](https://github.com/conventional-changelog/commitlint/assets/24919330/5e590a6d-27e0-44ea-9ce2-83ae7c700111)

</details>

<details>
<summary>

### References: Rules

</summary>

<img width="1274" alt="image" src="https://github.com/conventional-changelog/commitlint/assets/24919330/bb692a2c-cc6f-4c67-a650-adfcda5dfe7a">

</details>

JounQin commented:
I would recommend to use https://rspress.dev/

marcalexiei commented:
Pending discussion on https://github.com/conventional-changelog/commitlint/issues/3905#issuecomment-1947841896

marcalexiei commented:
I finished the main implementation of the docs using `vitepress`. 
I will keep updating this PR description with the relevant changes I made.

> [!TIP]
>
> If anyone want to take a look you can checkout this branch and try it locally and let me know what you think!
> To run the server simply use `yarn docs-dev`.

@escapedcat, @JounQin if you have time I would like to hear any feedback!

----

@escapedcat, I have a question about automatic installation:

When opening "Actions" pages I see that a "pages build and deployment" workflow exists:
https://github.com/conventional-changelog/commitlint/actions/workflows/pages/pages-build-deployment

However Inside `.github/workflows` I haven't find any workflow file for it.

Probably there is some kind of magic that I'm missing here 😅.  
There is any way I can see that code that runs this workflow so I can update it to use `vitepress`?

Just for reference this is the official `vitepress` deploy doc:
https://vitepress.dev/guide/deploy#github-pages

escapedcat commented:
That's all "Github pages magic":

<img width="818" alt="image" src="https://github.com/conventional-changelog/commitlint/assets/1016218/61d24d78-d458-48d9-b509-61b302aa50cc">

Currently everything in docs gets deployed.
We can adjust this if necessary.

marcalexiei commented:
Thanks for the info.
When the new doc is completed we have to set "Build and deployment → Source" to "GitHub Actions":

<img width="358" alt="image" src="https://github.com/conventional-changelog/commitlint/assets/24919330/c4effab9-fb52-4de8-8514-4200ac9201e5">

The new doc requires a build step before is ready for deploy.

Anyway there is more job to do before this 😀

---

One last question: do you want to add a copyright in the footer? 
In the current site there is none, however, since adding it is [trivial](https://vitepress.dev/reference/default-theme-footer#footer) let me know if you want it.

marcalexiei commented:
I completed the doc update. PR description is updated with all relevant changes.

I added also the `docs-deploy` workflow to automatically deploy the docs on every push on `master` branch.

Let me know if other changes are needed 😉.

----

After merge GitHub Pages repo settings should be updated as described in my [previous comment](https://github.com/conventional-changelog/commitlint/pull/3921#issuecomment-1951240303).

I never deployed pages on a custom domain so I'm not sure if other things should be update in the workflow / repository settings 😅.

JounQin commented on .github/CONTRIBUTING.md:
````diff
@@ -78,6 +78,18 @@ yarn test
+
+```sh
+yarn install
+yarn build-dev
````
`build-dev` seems not related to docs, how about `docs-dev`?

And I don't think `yarn install` is needed to be mentioned here?

marcalexiei replied:
> build-dev seems not related to docs, how about docs-dev?

I used referenced the wrong script: I meant to use `docs-dev`

> And I don't think yarn install is needed to be mentioned here?

I was thinking when someone clones and checkout the repo only to update documentation.

JounQin replied:
I think `yarn install` should be a separate section.

marcalexiei replied:
Right now every CONTRIBUTING.md section has a complete list of commands including 
- `yarn clean` 
- `yarn install`. 

E.g.:
- Testing
- Publishing a release
- Publish a `next` release (or i.e. patch release)

I don't see a valid reason to run `yarn clean` for working on the docs so I omitted it.

JounQin replied:
My point is that `install` is a one-time action, it should not be combined with `docs-dev`, HDYT?

marcalexiei replied:
How do you think about having something like this:

## Want to contribute? (h2)

with the following subsections (h3)

- Want to contribute?
- Commit Rules
- Environment setup → instructions **with** `yarn install`  
- Testing →  instructions **without** `yarn install`
- Documentation →   instructions **without** `yarn install`

marcalexiei replied:
(Publishing section are left as they are now)

JounQin replied:
I love these.

marcalexiei replied:
I'll update those sections with the next push.

JounQin commented on @commitlint/rules/src/index.test.ts:
```diff
@@ -22,29 +38,13 @@ test('rules export functions', () => {
 	expect(Object.keys(rules)).toEqual(expect.arrayContaining(results));
 });
-
-function _glob(pattern: string) {
```
Why these lines are changed?

marcalexiei replied:
Those functions were used only by the first test but they were declared athletes the end of the file.
I prefer to have function declaration before they are used.

JounQin replied:
I don't think it's related to document, and it's just personal taste, we should keep the diff as small as possible for easy reveiew experience. And function hoisted is absolute valid usage.

JounQin commented on docs/concepts/shareable-config.md:
````diff
@@ -3,16 +3,33 @@
+};
+```
+
+```ts [commitlint.config.ts]
````
Do we really need to add a TypeScript config example?

marcalexiei replied:
In the old docs some examples have the typescript code block some don't.

I think it's good to have some basic `typescript` examples:
- `reference/configuration` → "Configuration object example"
- `concepts/shareable-config` → Main section and "Relative config" 

Having those examples users can see where the correct types can be extracted.

IMHO , adding both `ts` and `js` examples in all scenarios is a bit redundant.
Anyway I have no problem in doing it.

JounQin replied:
Personally I think we don't need provide TypeScript example, we just need to point out that TypeScript config is supported.

marcalexiei replied:
What do you think about 
- removing all typescript examples 
- add a new section "Typescript configuration" in `reference/configuration` with reference to the type that should be used when writing the config

?

JounQin commented on docs/concepts/shareable-config.md:
````diff
@@ -24,37 +41,62 @@ This works recursively, enabling shareable configuration to extend on an indefin
-```js
-// commitlint.config.js
-module.exports = {
+```ts [commitlint.config.ts]
````
Like above, and I notice that some files have TypeScript config, some don't.

marcalexiei replied:
let's discuss this on this [thread](https://github.com/conventional-changelog/commitlint/pull/3921#discussion_r1500092570)

JounQin commented on docs/guides/getting-started.md:
````diff
@@ -0,0 +1,32 @@
+
+::: code-group
+
+```shell [npm]
````
```suggestion
```sh [npm]
```

Please use consistent language tag.

JounQin commented on docs/reference/configuration.md:
```diff
@@ -1,22 +1,43 @@
 # Configuration
 
-`@commitlint/cli` picks up configuration from `./commitlint.config.js`.
+## Config via file
```
Can we add a link to `cosmiconfig`'s documentation?

marcalexiei replied:
At the moment cosmiconfig@9 is out and we are using cosmiconfig@8.

After a quick search I wasn't able to find a v8 documentation link. Do you have one?

----

Out of curiosity, what is the reason to add a link to `cosmiconfig` documentation?
A user trying to setup commitlint shouldn't be affected by how `cosmiconfig` is configurable because is out of his control.
Right now `searchPlaces` are explicated on commitlint code:

https://github.com/conventional-changelog/commitlint/blob/0a583d362ae4afdfa627c4e72d5f30146b245a9f/%40commitlint/load/src/utils/load-config.ts#L40-L60

JounQin replied:
https://github.com/cosmiconfig/cosmiconfig/tree/v8.2.0

I think it's a good practice/respect to mention how config is resolved under the hood, and as you mentioned, cosmiconfig v9 is just released, we'll upgrade later to support `.config` dir, which is of course powdered by cosmiconfig.

---

IMO, we should not change `searchPlaces`, and there is no need, we should just follow the standard.

marcalexiei replied:
Ok, sounds good to me. I'll add the link after the list of resolved files.

marcalexiei replied:
I'll add it in the next push

JounQin commented on docs/reference/configuration.md:
```diff
@@ -131,9 +150,11 @@ const Configuration: UserConfig = {
   },
 };
 
-module.exports = Configuration;
```
We should mention that commonjs and ESM are both supported.

marcalexiei replied:
I've added a note alert under the code group

JounQin commented on docs/reference/configuration.md:
````diff
@@ -142,104 +163,113 @@ that can be resolved by the node resolve algorithm.
+::: code-group
+
+```js [commitlint.config.js]
+export default {
````
The jsdoc bases typings are not consistent across the documents.

marcalexiei replied:
Sorry, but I haven't understand what do you mean. Could you please elaborate?

JounQin replied:
In `docs/concepts/shareable-config.md`

<img width="562" alt="image" src="https://github.com/conventional-changelog/commitlint/assets/8336744/13a3d3c9-9d07-49b4-b305-793d71315e58">

But there is no jsdoc here, I don't think jsdoc is needed for such simple cases and there was not neither previously.

JounQin commented on docs/reference/examples.md:
```diff
@@ -0,0 +1,22 @@
+      "parserOpts": {
+        "issuePrefixes": ["PROJ-"],
+      },
+    },
```
Remove these trailing commas and add a `prettier-ignoreo` comment for https://github.com/prettier/prettier/issues/15956.

marcalexiei replied:
I added them manually by accident. `prettier` doesn't add them

JounQin replied:
Oh, yes, we haven't upgraded to prettier v3 yet.

JounQin commented on docs/reference/rules-configuration.md:
````diff
@@ -0,0 +1,33 @@
+## Plain array
+
+```js
+"rules": {
````
This seems not js.

marcalexiei replied:
To be consistent in all snippets I changed entire page to use code group with js highlighting the rules configuration:

<img width="734" alt="image" src="https://github.com/conventional-changelog/commitlint/assets/24919330/a3114a22-71fe-4dd1-86f2-64161a830eee">

JounQin commented on docs/guides/getting-started.md:
```diff
@@ -0,0 +1,32 @@
+
+## Install
+
+Install `commitlint` and a `commitlint-config-*` of your choice as devDependency and
```
```suggestion
Install `commitlint` and a `commitlint-config-*`/`@commitlint/config-*` of your choice as devDependency and
```

marcalexiei commented:
I have applied all requested changes. Let me know if other changes are needed.

JounQin commented on docs/api/lint.md:
````diff
@@ -0,0 +1,146 @@
+## Basic Example
+
+```js
+const lint = require('@commitlint/lint').default;
````
It's ESM only now. Many js config examples are affected.

marcalexiei replied:
My bad, I only searched for module.exports and I didn’t notice these examples when splitting reference-api file.

marcalexiei replied:
Now all `api/*` examples use `import` and `await`.

Additional changes:
- Each examples has been moved in a separate code snippet 
  (to simply copy-paste using the button provided by the code-block UI)
- Each code snippet has a title explaining what the snippet does

JounQin commented on docs/api/lint.md:
```diff
@@ -0,0 +1,146 @@
+  extends: ['@commitlint/config-conventional'],
+};
+
+load(CONFIG)
```
I'd prefer top level await.

marcalexiei replied:
Refer to this [comment](https://github.com/conventional-changelog/commitlint/pull/3921#discussion_r1500981209)

marcalexiei commented:
Thanks for your review @JounQin!

marcalexiei commented:
@escapedcat, if you don't have changes that needs to be applied, new documentation is ready for merge.

marcalexiei commented:
I just noticed that were many link pointing to the documentation with a `conventional-changelog.github.io/commitlint` domain.

I update all of them pointing to new the docs pages paths and to use `commitlint.js.org` domain.
I also removed some doc from single packages `README.md` and added reference to the docs pages and the associated `md` file used to generate it.

escapedcat commented:
Merged. Exciting. Let's hope for the best 🤞

escapedcat commented:
Had to update the branch-name, we're still on `master` :P  
Looks great:
https://commitlint.js.org/

Thanks everyone! You keep this alive!

escapedcat commented:
@JounQin would create a new major release tomorrow if you're good with that.  
I would expect some incoming issues/feedback :P

JounQin commented:
I would be glad for fixing any issues. 😁
